Tools/Cli-LoRa-Device-Provisioning/IoTDeviceHelper.cs

```diff
@@ -0,0 +1,1113 @@
+                twin.Properties.Desired[TwinProperty.RX2DataRate] = ValidationHelper.SetStringTwinProperty(opts.Rx2DataRate);
+
+            if (!string.IsNullOrEmpty(opts.Rx1DrOffset))
+                twin.Properties.Desired[TwinProperty.RX1DROffset] = ValidationHelper.SetUIntTwinProperty(opts.Rx1DrOffset);
+
+            if (!string.IsNullOrEmpty(opts.Supports32BitFCnt))
+                twin.Properties.Desired[TwinProperty.Supports32BitFCnt] = ValidationHelper.SetBoolTwinProperty(opts.Supports32BitFCnt);
+
+            return twin;
+        }
+
+        public async Task<bool> WriteDeviceTwin(Twin twin, string devEui, ConfigurationHelper configurationHelper, bool isNewDevice)
+        {
+            var device = new Device(devEui);
+            BulkRegistryOperationResult result;
+
+            Console.WriteLine();
+            Console.WriteLine($"Writing device {devEui} twin to IoT Hub...");
+
+            // Add new device
+            if (isNewDevice)
+            {
+                try
+                {
+                    result = await configurationHelper.RegistryManager.AddDeviceWithTwinAsync(device, twin);
+                }
+                catch (Exception ex)
+                {
+                    StatusConsole.WriteLine(MessageType.Error, ex.Message);
+                    return false;
+                }
+
+                if (result.IsSuccessful)
+                {
+                    Console.WriteLine($"Success!");
+                }
+                else
+                {
+                    Console.WriteLine($"Error adding device:");
+                    foreach (var error in result.Errors)
+                    {
+                        Console.WriteLine($"Device Id: {error.DeviceId}, Code: {error.ErrorCode}, Error: {error.ErrorStatus}");
+                    }
+
+                    return false;
+                }
+            }
+
+            // Update existing device
+            else
+            {
+                try
+                {
+                    await configurationHelper.RegistryManager.UpdateTwinAsync(twin.DeviceId, twin, twin.ETag);
+                }
+                catch (Exception ex)
+                {
+                    StatusConsole.WriteLine(MessageType.Error, ex.Message);
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public async Task<bool> QueryDevices(ConfigurationHelper configurationHelper, int page, int total)
+        {
+            var count = 0;
+            IEnumerable<string> currentPage;
+            string totalString = (total == -1) ? "all" : total.ToString();
+
+            Console.WriteLine();
+            Console.WriteLine($"Listing devices...");
+            Console.WriteLine($"Page: {page}, Total: {totalString}");
+            Console.WriteLine();
+
+            var query = configurationHelper.RegistryManager.CreateQuery("SELECT * FROM devices", page);
```

Comment:
page is not validated. What if it is negative?

Reply:
Done!